types/rlp_unmarshal.go

```diff
@@ -376,44 +384,62 @@ func (t *Transaction) unmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 		return err
 	}
 
-	if len(elems) < 9 {
-		return fmt.Errorf("incorrect number of elements to decode transaction, expected 9 but found %d", len(elems))
+	getElem := func() *fastrlp.Value {
+		val := elems[0]
+		elems = elems[1:]
+
+		return val
+	}
+
+	var num int
+
+	switch t.Type {
+	case LegacyTx:
+		num = 9
+	case DynamicFeeTx, StateTx:
+		num = 10
+	default:
+		return fmt.Errorf("transaction type %d not found", t.Type)
+	}
+
+	if numElems := len(elems); numElems != num {
+		return fmt.Errorf("incorrect number of transaction elements, expected %d but found %d", num, numElems)
 	}
 
 	p.Hash(t.Hash[:0], v)
 
 	// nonce
-	if t.Nonce, err = elems[0].GetUint64(); err != nil {
+	if t.Nonce, err = getElem().GetUint64(); err != nil {
 		return err
 	}
 
 	if t.Type == DynamicFeeTx {
 		// gasFeeCap
 		t.GasFeeCap = new(big.Int)
-		if err = elems[10].GetBigInt(t.GasFeeCap); err != nil {
+		if err = getElem().GetBigInt(t.GasFeeCap); err != nil {
```

Comment:
Correct me if I'm wrong, but this getElem will not return the correct values if this t.Type == DynamicFeeTx condition is met, because it goes from the first element in array and decreases the array size by each call, and we were extracting the 10th and 11th index from the array before.

Reply:
It returns elements with the same indecies basically. Dynamic base fee tx has these two fields instead of the gas price one (according to eth spec). Could you please provide more details if I'm missing something? Maybe a specific test case when it would not work?

Reply:
Yeah, it confused me, because I am looking the code comparison, where on branch `feature/EVM-262-EIP-1559` on which this PR is created, you are taking these `GasFeeCap` and `GasTipCap` from 10th and 11th index, but I didn't see the change in `MarshalRLP` when you insert this data in array.

But I see you changed in some earlier PR the `MarshalRLP` function, that got merged to `v3-parity`, so those fields get marshaled on 1st and 2nd index if it is a `DynamicFeeTx`, so it looks like everything is fine in this PR.

Do you need to rebase this PR to `v3-parity` perhaps?

Reply:
Yea, I've changed ordering according to another eip so it should be: (gas price) OR (base tip cap, base fee cap). Both marshaling and unmarshaling functions were updated accordingly. This branch should be against `feature/EVM-262-EIP-1559` since that one is the parent branch that should be reviewed after everything related to 1559 is done.